CIP-0020/CIP-0020.md

```diff
@@ -0,0 +1,102 @@
+         }
+}
+```
+
+
+
+Example for a single message/comment/memo:
+``` json
+{ 
+  "674":
+         { "msg": 
+                  [ 
+                    "This is a comment for the transaction xyz, thank you very much!"
+                  ]
+         }
+}
+```
+
+Example for multiple messages/comments/memos:
+``` json
+{ 
+  "674":
+         { "msg": 
+                  [ 
+                    "Invoice-No: 1234567890",
+                    "Customer-No: 555-1234",
+                    "P.S.: i will shop again at your store :-)"
+                  ]
+         }
+}
+```
+Example above in **Daedalus** currently (could be improved if CIP is implemented):
+![image](https://user-images.githubusercontent.com/47434720/121822100-85b38a80-cc9d-11eb-9d13-1869746a69b2.png)
+
+**Cardanoscan.io**, **Adastat.net** and other tools implemented it already, to show messages along transactions:
+![image](https://user-images.githubusercontent.com/47434720/124379245-1f2af680-dcb6-11eb-97b7-10f70d840e88.png)
+![image](https://user-images.githubusercontent.com/47434720/124381343-3ff94900-dcc2-11eb-8d03-8fbacd3322b0.png)
+
+**ccwallet.io** has added it with a message field on the sending-page, and shows it also on the transactions-page:
+![image](https://user-images.githubusercontent.com/47434720/127367420-b360972d-c6e0-4002-865e-df070904bd30.png)
+![image](https://user-images.githubusercontent.com/47434720/127367228-339ac059-007a-40fd-a6c0-97f890e93964.png)
+![image](https://user-images.githubusercontent.com/47434720/127368912-c85dc9f0-6ee3-4cc1-a24b-9716a20f27d3.png)
+
+
+# Rationale
+
+This design is simple, so many tools on the cardano blockchain can implement it easily. The array type was choosen to have consistency, no need to switch between a string or
+an array format, or testing against a string or array format. Updates in the future are possible, like adding a versioning key `"ver":`, adding a key `"utxo":` to provide specific data for every tx-out#idx in the transaction, making subarrays in the message-strings, etc. But for now, we need a common agreement to provide general messages/comments/memos with this CIP. The starting design war choosen as simple as possible to keep the additional transaction fees as low as possible.
+
+## Wallet Implementation
+
+Would be a good idea to hide the message/comment/note behind a "show unmoderated content" button/drop-down. Like the Metadata display on the Cardano Explorer. Also, it should be displayed as plain-text non-clickable. To enhance security further, URLs could be automatically deleted or hidden from such comments, to not welcome bad actors with phishing attempts. Another solution to start with would be to really limit the character space for display in Wallets, like limiting it to `a-zA-z0-9` and a handful of special chars like `+-_#()[]:` without a `.<>"/\` chars, so a domain or html code would not work. Last points are worth for discussions of course, because it would also filter out unicode.
+
```

Comment:
Handling ill-formed 674 metadata: It is up to the wallet-/display-/receiver-implementor to parse and check the provided metadata. As for the current state, its not possible to have the same label "674" more than once in a cardano transaction. So a check about that can be ignored at the moment. This CIP provides the correct implementation format, the parsing should search for the "674" metadata label and the "msg" key underneath it. There should also be a check, that the provided data within that "msg" key is an array. All other implementations like a missing "msg" key, or a single string instead of an array, should be marked by the display-implementor as "invalid". Additional keys within the "674" label should not affect the parsing of the "msg" key. As written above, we will likely see more entries here in the future like a "version" key for example, so additional keys should not harm the parsing of the "msg" key. The implementation format in this CIP should be the ground base for transactions messages/comments/memos and should be respected by creator-implemantations as well as in receiver-implementations.